arch/risc-v/src/mpfs/mpfs_sdio.c

```diff
@@ -0,0 +1,177 @@
+ *
+ * Returned Values:
+ *   A reference to an SDIO interface structure.  NULL is returned on
+ *   failures.
+ *
+ ****************************************************************************/
+
+struct sdio_dev_s *sdio_initialize(int slotno)
+{
+  switch (slotno)
+    {
+    case 0:
+#ifdef CONFIG_MPFS_EMMCSD
+      return mpfs_emmcsd_sdio_initialize(slotno);
+#endif
+
+    case 1:
+#ifdef CONFIG_MPFS_COREMMC
+      return mpfs_coremmc_sdio_initialize(slotno);
+#endif
+
+    default:
+      break;
+    }
+
+  mcerr("sdio slot number %d not supported!\n", slotno);
+  return NULL;
+}
+
+/****************************************************************************
+ * Name: sdio_mediachange
+ *
+ * Description:
+ *   Called by board-specific logic -- possible from an interrupt handler --
+ *   in order to signal to the driver that a card has been inserted or
+ *   removed from the slot
+ *
+ * Input Parameters:
+ *   dev        - An instance of the SDIO driver device state structure.
+ *   cardinslot - true is a card has been detected in the slot; false if a
+ *                card has been removed from the slot.  Only transitions
+ *                (inserted->removed or removed->inserted should be reported)
+ *
+ * Returned Value:
+ *   None
+ *
+ ****************************************************************************/
+
+void sdio_mediachange(struct sdio_dev_s *dev, bool cardinslot)
+{
+  struct mpfs_dev_s *priv = (struct mpfs_dev_s *)dev;
+
+  if (!dev)
+    {
+      mcerr("sdio device not found\n");
+      return;
+    }
+
+  switch (priv->hw_base)
+    {
+#ifdef CONFIG_MPFS_EMMCSD
+    case MPFS_EMMC_SD_BASE:
+      mpfs_emmcsd_sdio_mediachange(dev, cardinslot);
+      return;
+#endif
+
+#ifdef CONFIG_MPFS_COREMMC
+    case CONFIG_MPFS_COREMMC_BASE:
+      mpfs_coremmc_sdio_mediachange(dev, cardinslot);
+      return;
+#endif
+
+    default:
+      break;
+    }
+
+  mcerr("Invalid sdio base address\n");
+}
+
+/****************************************************************************
+ * Name: sdio_wrprotect
+ *
+ * Description:
+ *   Called by board-specific logic to report if the card in the slot is
+ *   mechanically write protected.
+ *
+ * Input Parameters:
+ *   dev       - An instance of the SDIO driver device state structure.
+ *   wrprotect - true is a card is writeprotected.
+ *
+ * Returned Value:
+ *   None
+ *
+ ****************************************************************************/
+
+void sdio_wrprotect(struct sdio_dev_s *dev, bool wrprotect)
+{
+  struct mpfs_dev_s *priv = (struct mpfs_dev_s *)dev;
+
+  if (!dev)
+    {
+      mcerr("sdio device not found\n");
+      return;
+    }
+
+  switch (priv->hw_base)
+    {
+#ifdef CONFIG_MPFS_EMMCSD
+    case MPFS_EMMC_SD_BASE:
+      mpfs_emmcsd_sdio_wrprotect(dev, wrprotect);
+      return;
+#endif
+
+#ifdef CONFIG_MPFS_COREMMC
+    case CONFIG_MPFS_COREMMC_BASE:
+      mpfs_coremmc_sdio_wrprotect(dev, wrprotect);
+      return;
+#endif
+
+    default:
```

Comment:
Please add a message error in case the hw_base is invalid.

Reply:
The error message is right below the switch statement block, where the code execution ends up if the function did not return from one of the known hw_base cases.

https://github.com/apache/nuttx/blob/47c85ab0b35fbe07ff1599710a9be2ec2b7d8505/arch/risc-v/src/mpfs/mpfs_sdio.c#L169-L177

But I can move it into the "default:" case if it is better that way.
Or did you mean something else?

Reply:
Well.. maybe better to move it to avoid misinterpretations